[Discover/Lens] Unify field popover

jughosta opened the pull request:
Closes https://github.com/elastic/kibana/issues/140363

## Summary

This PR creates a new shared component `<FieldPopover .../>` in UnifiedFieldList plugin and integrates it back to Discover and Lens.

- [x] Integrate into Discover
- [x] Integrate into Lens
- [x] New shared component for "Visualize" button
- [x] Tests
- [x] Update docs

Discover:
<img width="400" alt="Screenshot 2022-09-30 at 18 25 43" src="https://user-images.githubusercontent.com/1415710/193315055-c99ea89f-ce0f-4f4c-b8be-5f2ff4ce176d.png">


Lens:
<img width="400" alt="Screenshot 2022-09-30 at 18 25 24" src="https://user-images.githubusercontent.com/1415710/193315099-35990310-1624-46af-bc8a-df6d09df6653.png">




### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

jughosta commented on src/plugins/discover/public/application/main/components/chart/discover_chart.tsx:
```diff
@@ -18,16 +18,18 @@ import {
+import {
+  getVisualizeInformation,
+  triggerVisualizeActions,
+} from '@kbn/unified-field-list-plugin/public';
```
@davismcphee In this PR I extracted `<FieldVisualizeButton />` to unified-field-list and moved these functions inside unified-field-list-plugin too. I guess you would also need them in unified-histogram, right? What could be a better solution for organising it?

davismcphee replied:
Good question, I haven't got to these yet in my refactoring so I haven't thought much about it, but I'll look into it and let you know what I think.

jughosta commented:
While working on this topic I added "+" button to the popover also on Discover page. And for consistency with other places I switched the icon from `plusInCircle` to `plusInCircleFilled`. What icon should we use? Wdyt @andreadelrio @ninoslavmiskovic 

Before:
<img width="300" alt="Screenshot 2022-09-29 at 16 59 49" src="https://user-images.githubusercontent.com/1415710/193067305-b05c37ff-8a2d-4c4a-b5af-99bdf1c3bf96.png">
<img width="300" alt="Screenshot 2022-09-29 at 17 00 08" src="https://user-images.githubusercontent.com/1415710/193067310-ad3703a1-7ddf-4e41-a29c-1ef60d95a07a.png">

After:
<img width="300" alt="Screenshot 2022-09-29 at 16 58 59" src="https://user-images.githubusercontent.com/1415710/193067379-56faba0a-cdce-4946-9710-d4c6a0b4e32a.png">
<img width="300" alt="Screenshot 2022-09-29 at 16 58 32" src="https://user-images.githubusercontent.com/1415710/193067371-5367b294-c305-4eb6-9e57-545116be6bab.png">

jughosta commented on src/plugins/unified_field_list/public/components/field_visualize_button/field_visualize_button.tsx:
```diff
@@ -7,29 +7,48 @@
 
-export const DiscoverFieldVisualize: React.FC<Props> = React.memo(
-  ({ field, dataView, contextualFields, trackUiMetric, multiFields }) => {
+export const FieldVisualizeButton: React.FC<FieldVisualizeButtonProps> = React.memo(
```
Hi @stratoula ! In this PR I am moving "Visualize" button and the related code to unifiedFieldList plugin. And I am wondering if we could also move "Explore in Discover" button here as well. I guess I would then have to convert it into an action and trigger in the same way as we do for "Visualize" button.  Would it make sense to do so?

stratoula replied:
Hmm, yes we should def move it on the unified plugin. The reasons we used href here(using the discover locator here) are:
- it is in sync with the same link that exists on the topNav of Lens
- we open it on a new tab

Can we keep the same functionality here without adding a new trigger action?

jughosta replied:
@stratoula Yes, with a trigger we could also open in a new tab and it could be similar to [lens/public/trigger_actions/open_in_discover_action.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/lens/public/trigger_actions/open_in_discover_action.ts) and [unified_field_list/public/components/field_visualize_button/field_visualize_button_inner.tsx](https://github.com/jughosta/kibana/blob/a66193049cf1b7bed49874eafc6c16eca073bd1e/src/plugins/unified_field_list/public/components/field_visualize_button/field_visualize_button_inner.tsx#L36) (at least from what I understand about triggers). 
On the other hand, a direct call to discover locator without a trigger would be problematic from unifiedFieldList unless we add it as a dependency.

stratoula replied:
Ok then, if we can have the same functionality let's do it with a uiAction instead :)

stratoula replied:
@jughosta it could also be a prop right? I mean whichever app wants to use the unified field list and wants to display the button can give the props of the button (label, action etc) so it depends on the application. This could possibly make the field list more flexible (?).

jughosta replied:
@stratoula Yes, that might work, thanks! Created a follow up ticket: https://github.com/elastic/kibana/issues/142361 
Anyway, the current implementation is also simple and good so it can definitely wait :)

jughosta commented:
@elasticmachine merge upstream

jughosta commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

jughosta commented on x-pack/plugins/lens/public/data_views_service/loader.ts:
```diff
@@ -47,6 +47,9 @@ export function convertDataViewIntoLensIndexPattern(
         meta: dataView.metaFields.includes(field.name),
         esTypes: field.esTypes,
         scripted: field.scripted,
+        isMapped: field.isMapped,
+        customLabel: field.customLabel,
+        runtimeField: field.runtimeField,
```
Not sure if this is okay to extend. What was the reason to construct `IndexPatternField` instead of `DataViewField`?

MichaelMarcialis commented:
@yiyangliu9286: Seeing as this is related to the work you're preparing to do on the unification of the Discover/Lens field lists, did you want to give this PR a review? If not, let me know and I can give it a quick review. Thanks!

yiyangliu9286 commented:
> @yiyangliu9286: Seeing as this is related to the work you're preparing to do on the unification of the Discover/Lens field lists, did you want to give this PR a review? If not, let me know and I can give it a quick review. Thanks!

Thanks for tagging @MichaelMarcialis! I am learning how to setup local environment at the moment so it might take longer for me to do the actual PR review. If you don't mind to do a quick review it would be great. I'll provide some design input for above questions.

yiyangliu9286 commented:
> While working on this topic I added "+" button to the popover also on Discover page. And for consistency with other places I switched the icon from `plusInCircle` to `plusInCircleFilled`. What icon should we use? Wdyt @andreadelrio @ninoslavmiskovic
> 
> Before: <img alt="Screenshot 2022-09-29 at 16 59 49" width="300" src="https://user-images.githubusercontent.com/1415710/193067305-b05c37ff-8a2d-4c4a-b5af-99bdf1c3bf96.png"> <img alt="Screenshot 2022-09-29 at 17 00 08" width="300" src="https://user-images.githubusercontent.com/1415710/193067310-ad3703a1-7ddf-4e41-a29c-1ef60d95a07a.png">
> 
> After: <img alt="Screenshot 2022-09-29 at 16 58 59" width="300" src="https://user-images.githubusercontent.com/1415710/193067379-56faba0a-cdce-4946-9710-d4c6a0b4e32a.png"> <img alt="Screenshot 2022-09-29 at 16 58 32" width="300" src="https://user-images.githubusercontent.com/1415710/193067371-5367b294-c305-4eb6-9e57-545116be6bab.png">

@jughosta imo it might be better/more consistent to use `plusInCircle` instead `plusInCircleFilled` (same with Lens) to align with the list of top values "+" icon within this Popover and also to align with all the control icons on the right hand side (filter, edit, delete etc.) - For the hover over "+" icon on the side bar & Multi fields "+" let's keep it as it because they are hovering interaction (on a non-white background) and we should reconsider in the future. 

<img width="581" alt="Discover popover" src="https://user-images.githubusercontent.com/81987435/193896277-17686465-5093-4283-9115-e5682efca64e.png">

stratoula reviewed (approved):
+1 to use plusInCircle instead as Lens does. 

Except from that everything works as expected (I was mostly focused on the Lens editor) so I am approving :)

kertal reviewed (approved):
+1 for the Lens way! Code LGTM, tested locally in Safari, Firefox, Chrome, works as expected. Actually I like the consistency with Lens a lot 👍

jughosta commented:
Updated to `plusInCircle` icon, thanks for your feedback! 🙌

jughosta commented:
@yiyangliu9286 @MichaelMarcialis A demo deployment is also available under [this link](https://jughosta-pr-141787-140363-unified-popover.kbndev.co/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-2d,to:now))&_a=(columns:!(),filters:!(),index:af31763a-b03b-4518-8717-bcef30502926,interval:auto,query:(language:kuery,query:''),sort:!(!('@timestamp',desc)))) via elastic/changeme
Thanks!

MichaelMarcialis commented on src/plugins/unified_field_list/public/components/field_popover/field_popover_header.tsx:
```diff
@@ -0,0 +1,161 @@
+import type { DataViewField } from '@kbn/data-views-plugin/common';
+import type { AddFieldFilterHandler } from '../../types';
+
+const titleStyles = css`
+  text-transform: none;
+`;
```
Out of curiosity, is this style necessary? I'm thinking this is probably a straggler override style from when the previous EUI theme was in use and forced uppercase on all popover headers. However, the current theme no longer does this. If no longer necessary, I imagine it can be removed.

jughosta replied:
Removed styles, thanks @MichaelMarcialis !

MichaelMarcialis reviewed (approved):
Hey, @jughosta! I left one small comment for your review, but nothing super important. Assuming it gets addressed, I'll go ahead and approve this now. Thanks!
